---
title: "Differential Expression Analysis"
output: 
  html_document:
    toc: true
    toc_float:
      collapsed: false
    fig_width: 5
editor_options: 
  chunk_output_type: console
---

<style type="text/css">

blockquote {
  background: #ECF8FF;
  border-left: 10px solid #3989CB;
  margin: 1.5em 10px;
  padding: 0.5em 10px;
  font-size: 14px;
}

h1 { 
  font-size: 25px;
  margin-top: 1.5cm;
  margin-bottom: 0.5cm;
}
h2 { 
  font-size: 18px;
  margin-top: 1cm;
  margin-bottom: 0.5cm;
}
h3 {
  font-size: 14px;
  margin-top: 1cm;
  margin-bottom: 0.5cm;
}

table.answer, td.answer {
   border: 0px;
   background: #BCE0C0;
   padding: 10px;
   width: 100%;
}

div.answer { display: none;}

</style>


```{r setup, include=F}
knitr::opts_chunk$set(echo = TRUE, tidy=F, eval=T, cache=F)
qn <- sn <- 0
```

Setup and load data 
===================

This exercise will use unpublished RNA-Seq data (gene counts) from cultures of *Saccharomyces cerevisiae* before and after pheramone treatment, and should be in the `data/` subdirectory in your R working directory. 

The `gene_counts_glucose.txt` dataset corresponds to glucose as the carbon source, whereas the `gene_counts_ethanol.txt` dataset used ethanol. Each dataset includes 6 samples prior to pheramone treatment, and several samples (24 for glucose, 15 for ethanol) after treatment.

`r sn<-sn+1; paste(sn,". ", sep="")` Start R and use `setwd()` (or Session > Set Working Directory > Choose Directory... from the top menu bar) to change the working directory to the `IDA_exercise/` directory, which should contain a directory named `data/`.

`r sn<-sn+1; paste(sn,". ", sep="")` We will use the following R packages in this exercise:
```{r, message=FALSE, warning=FALSE, results='hide'}
#library(knitr)
#library(topGO)
#library(biomaRt)
#library(piano)
#library(NMF)
#library(org.Mm.eg.db)
#library(Rgraphviz)
library(edgeR)
```

If you are not able to load any of these packages, try to install them, either using `biocLite()` (first you need to run `source("http://www.bioconductor.org/biocLite.R")`) or `install.packages()`. If something fails, try to understand the error message and fix it. If you get stuck, ask for help :)  

`r sn<-sn+1; paste(sn,". ", sep="")` Begin by loading the count data corresponding to the experiment of interest (glucose or ethanol conditions):
```{r loadcounts}
# load ethanol data
count.data <- read.delim("data/gene_counts_ethanol.txt", stringsAsFactors=F, row.names=1)

# load glucose data
#count.data <- read.delim("data/gene_counts_glucose.txt", stringsAsFactors=F, row.names=1)

# preview the first few rows of the data
head(count.data)
```


> **Question `r qn<-qn+1;qn`:** _____?

<div class="answer"><table class="answer"><tr><td class="answer">
**Answer:** 
```{r}
#sum(diffExpRes$FDR<0.001)
```
</td></tr></table></div>



Differential expression analysis
================================

We will perform a differential expression (DE) analysis using the [edgeR](https://bioconductor.org/packages/release/bioc/html/edgeR.html) package. This exercise only uses/explains a little of edgeR, so it is strongly encouraged that you take a look at the [edgeR User Manual](https://bioconductor.org/packages/release/bioc/vignettes/edgeR/inst/doc/edgeRUsersGuide.pdf) for further information.

Another package that is commonly used for DE analysis is [DESeq2](https://bioconductor.org/packages/release/bioc/html/DESeq2.html), which has a very nice [User Guide](https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html). Although DESeq2 will not be covered in this exercise, we encourage you to check it out sometime.


edgeR
-----

We want to determine genes that are differentially expressed between two groups - in this case, between "untreated" and "pheramone-treated" samples, which are indicated by column names beginning with a "U" or "P", respectively. We can take a quick look at the column names to see what samples we have:
```{r}
colnames(count.data)
```


`r sn<-sn+1; paste(sn,". ", sep="")` Generate a factor specifying the group to which each sample belongs:
```{r}
sample.groups <- factor(startsWith(colnames(count.data), 'P'), labels=c('untreated','pheramone'))

# check that we set up the groups correctly
data.frame(sample=colnames(count.data), groups=sample.groups)
```


`r sn<-sn+1; paste(sn,". ", sep="")` Create a `DGEList` object from the `counts` dataframe and `sample.groups` factor:
```{r}
y <- DGEList(counts=count.data, group=sample.groups)
```


Before performing any DE analysis, we can take a look at similarities and differences among different samples using a multidimensional scaling (MDS) plot:
```{r}
plotMDS(y)
```


`r sn<-sn+1; paste(sn,". ", sep="")` Filter (remove) low-count genes from the count data:
```{r}
keep <- filterByExpr(y, min.count = 10)
y <- y[keep, , keep.lib.sizes=F]
```

The `filterByExpr` function will only keep genes that have at least `min.count` counts per million (CPM) in at least `X` samples, where `X` is the number of samples in the smallest group (the `untreated` group, in this case).

The 'keep.lib.sizes=F' argument specifies that we want to recalculate the sample library sizes after removing the low-count genes from the data.


`r sn<-sn+1; paste(sn,". ", sep="")` Calculate "normalization factors" for each sample, to normalize by RNA composition:
```{r}
y <- calcNormFactors(y)
```

Normalizing by RNA composition is done to address problems caused when the counts in a sample are dominated by a few highly expressed genes, which can give other genes in that sample artificially low count values. The normalization factors calculated using the `calcNormFactors` function are used to scale the counts to avoid such problems.

To view the normalization factors calculated for each sample (as well as library sizes), we can take a look at the `samples` attribute of our DGEList, `y`:
```{r}
y$samples
```


`r sn<-sn+1; paste(sn,". ", sep="")` Estimate the dispersion across all genes:
```{r}
y <- estimateDisp(y)
```

We can also view the dispersion estimates using the BCV plot:
```{r}
plotBCV(y)
```


`r sn<-sn+1; paste(sn,". ", sep="")` Now we can test for DE genes using the `exactTest` function:
```{r}
et <- exactTest(y)
topTags(et)  # inspect the top-DE genes
```

The table of DE results is saved in `et$table`, but it does not contain the FDR-adjusted p-values. We can add this using the `p.adjust` function:
```{r}
res.table <- et$table
res.table$FDR <- p.adjust(et$table$PValue, 'BH')  # BH = Benjamini-Hochberg method
```


`r sn<-sn+1; paste(sn,". ", sep="")` Export the results table to a .csv file:
```{r, eval=F}
write.csv(res.table, file = 'DE_results_EtOH.csv')
```



Session info
============
This page was generated using the following R session:
```{r, echo=F}
sessionInfo()
```

