---
title: "Differential Expression Analysis"
output: 
  html_document:
    toc: true
    toc_float:
      collapsed: false
    fig_width: 5
editor_options: 
  chunk_output_type: console
---

<style type="text/css">

blockquote {
  background: #ECF8FF;
  border-left: 10px solid #3989CB;
  margin: 1.5em 10px;
  padding: 0.5em 10px;
  font-size: 14px;
}

h1 { 
  font-size: 25px;
  margin-top: 1.5cm;
  margin-bottom: 0.5cm;
}
h2 { 
  font-size: 18px;
  margin-top: 1cm;
  margin-bottom: 0.5cm;
}
h3 {
  font-size: 14px;
  margin-top: 1cm;
  margin-bottom: 0.5cm;
}

table.answer, td.answer {
   border: 0px;
   background: #BCE0C0;
   padding: 10px;
   width: 100%;
}

div.answer { display: true;}

</style>


```{r setup, include=F}
knitr::opts_chunk$set(echo=T, tidy=F, eval=T, cache=F)
qn <- sn <- 0
```

R session setup
===============

This exercise will use unpublished (!) RNA-Seq data (gene counts) from cultures of *Saccharomyces cerevisiae* that have been treated with a pheromone, and are classified as either pheromone-affected and pheromone-unaffected. The count data files are located in the `data/gene_counts/` subdirectory in the `IDA_tutorial` directory.

The `gene_counts_glucose.txt` dataset corresponds to glucose as the carbon source, whereas the `gene_counts_ethanol.txt` dataset used ethanol. Each dataset includes 6 pheromone-affected samples, and several pheromone-unaffected samples (24 for glucose, 15 for ethanol).

`r sn<-sn+1; paste(sn,". ", sep="")` Start R and use `setwd()` (or Session > Set Working Directory > Choose Directory... from the top menu bar) to change the working directory to the `IDA_tutorial/` directory, which should contain a subdirectory named `data/`.

`r sn<-sn+1; paste(sn,". ", sep="")` Load the `edgeR` package that will be used in this exercise:
```{r, message=FALSE, warning=FALSE, results='hide'}
library(edgeR)
```

If you are unable to load the package, try to install it using the following commands:
```{r, eval=F}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install('edgeR')
```

If something fails, try to understand the error message and fix it. If you get stuck, ask for help :)  


Load data
=========

`r sn<-sn+1; paste(sn,". ", sep="")` Begin by loading the count data corresponding to the experiment of interest (glucose or ethanol conditions):
```{r}
count.data <- read.delim('data/gene_counts/gene_counts_glucose.txt', stringsAsFactors=F, row.names=1)
# count.data <- read.delim('data/gene_counts/gene_counts_ethanol.txt', stringsAsFactors=F, row.names=1)

# preview the first few rows of the data
head(count.data)
```


> **Question `r qn<-qn+1;qn`:** How many genes are in the dataset?

<div class="answer"><table class="answer"><tr><td class="answer">
**Answer:** 
```{r}
nrow(count.data)
```
</td></tr></table></div>



Differential expression analysis
================================

We will perform a differential expression (DE) analysis using the [edgeR](https://bioconductor.org/packages/release/bioc/html/edgeR.html) package. This exercise only uses/explains a little of edgeR, so it is strongly encouraged that you take a look at the [edgeR User's Guide](https://bioconductor.org/packages/release/bioc/vignettes/edgeR/inst/doc/edgeRUsersGuide.pdf) for further information.

Another package that is commonly used for DE analysis is [DESeq2](https://bioconductor.org/packages/release/bioc/html/DESeq2.html), which also has a very nice [User Guide](https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html). Although DESeq2 will not be covered in this exercise, we encourage you to check it out sometime.


Data setup and exploration
--------------------------

We want to determine genes that are differentially expressed between two groups of samples - in this case, between "pheromone-unaffected" and "pheromone-affected" samples, which we have indicated by column names beginning with a "U" or "P", respectively. We can take a quick look at the column names to see what samples we have:
```{r}
colnames(count.data)
```


`r sn<-sn+1; paste(sn,". ", sep="")` Generate a factor specifying the group to which each sample belongs:
```{r}
sample.groups <- factor(startsWith(colnames(count.data), 'P'), labels=c('unaffected','pheromone'))

# check that we set up the groups correctly
data.frame(sample=colnames(count.data), groups=sample.groups)
```


`r sn<-sn+1; paste(sn,". ", sep="")` Create a `DGEList` object from the `count.data` dataframe and `sample.groups` factor:
```{r}
y <- DGEList(counts=count.data, group=sample.groups)
```


Before performing any DE analysis, we can take a look at similarities and differences among different samples using a multidimensional scaling (MDS) plot:
```{r}
plotMDS(y, cex=0.8)  # "cex" sets the label font size
```


> **Question `r qn<-qn+1;qn`:** What is MDS, and how does it compare/contrast with related techniques, such as principal component analysis (PCA)?

<div class="answer"><table class="answer"><tr><td class="answer">
**Answer:** 
Like PCA, MDS is a dimensionality reduction technique that allows the visualization of similarities among individal items in a dataset. Unlike PCA, MDS is non-linear transformation, and involves the minimization of a loss function to arrive at the final set of coordinates.
</td></tr></table></div>


> **Question `r qn<-qn+1;qn`:** Does it look like our samples are grouping/separating well based on pheromone status? Do you notice any other patterns in the data?

<div class="answer"><table class="answer"><tr><td class="answer">
**Answer:** 
Yes, the pheromone-unaffected (U) and pheromone-affected (P) samples clearly separate on the x-axis, indicating a substantial difference in transcriptomic signatures between the two conditions.

Another feature is a greater spread/difference among U samples than among P samples, but this could in part be due to a larger number of U samples.

Since the U vs. P difference is so great, it can be hard to pick up on any other patterns that may be present. To investigate this further, we can take a look at only the U samples:
```{r}
plotMDS(DGEList(counts=count.data[,7:ncol(count.data)]), cex=0.8)
```

It looks like there may be some further grouping among the unaffected (U) samples...
</td></tr></table></div>


Filtering and normalization
---------------------------

Often times RNA-Seq data will contain low-count genes that are not expressed, or expressed at very low levels in many samples. The data for these genes is essentially noise that does not provide any meaningful information, and should therefore be removed prior to DE analysis.

Although there is no single "correct" approach to remove low-count genes, a common technique involves removing those that are below some threshold in a given fraction of the samples.

`r sn<-sn+1; paste(sn,". ", sep="")` Filter (remove) low-count genes from the count data:
```{r}
keep <- filterByExpr(y, min.count = 10)
y <- y[keep, , keep.lib.sizes=F]
```

The `filterByExpr` function will only keep genes that have at least `min.count` counts per million (CPM) in at least `n` samples, where `n` is the number of samples in the smallest group (the `pheromone` group, in this case, where `n = 6`).

The `keep.lib.sizes=F` argument specifies that we want to recalculate the sample library sizes after removing the low-count genes from the data.


<br>The next step is to normalize the data by RNA composition, which is done to address problems caused when the counts in a sample are dominated by a few highly expressed genes. This can cause other genes in the sample to exhibit artificially low count values. The normalization factors calculated using the `calcNormFactors` function are used to scale the counts to avoid such problems.

`r sn<-sn+1; paste(sn,". ", sep="")` Calculate the normalization factors for each sample:
```{r}
y <- calcNormFactors(y)
```

To view the normalization factors calculated for each sample (as well as library sizes), we can take a look at the `samples` attribute of our DGEList, `y`:
```{r}
y$samples
```


Dispersion estimation
---------------------

`r sn<-sn+1; paste(sn,". ", sep="")` Estimate the dispersion across all genes:
```{r, results="hide"}
y <- estimateDisp(y)
```

We can also view the dispersion estimates using the BCV plot:
```{r}
plotBCV(y)
```


Test for differential expression
--------------------------------

`r sn<-sn+1; paste(sn,". ", sep="")` Now we can test for DE genes using the `exactTest` function:
```{r}
et <- exactTest(y)
topTags(et)  # inspect the top-DE genes
```


Export results
--------------

The table of DE results is saved in `et$table`, but it does not contain the FDR-adjusted p-values. We can add this using the `p.adjust` function:
```{r}
res.table <- et$table
res.table$FDR <- p.adjust(et$table$PValue, 'BH')  # BH = Benjamini-Hochberg method
```

`r sn<-sn+1; paste(sn,". ", sep="")` Export the results table to a .txt file:
```{r, eval=T}
# note: this command will overwrite the existing pre-generated DE_results_glucose.txt file
write.table(res.table, file = 'data/DE_results/DE_results_glucose.txt', quote=F, sep='\t')
```


Data with multiple variables
============================

In many cases, samples can be classified/grouped by more than one experimental variable/condition. For example, in a clinical trial, we may have unreated and treated subjects, as well as male and female. Often times we may want to look at differences between the different grouping types (untreated vs. treated, and male vs. female), or attempt to prevent factors associated with one variable from influencing our analysis of another. For example, if we are only interested in the effect of the drug treatment, we don't want differences in subject gender to bias the results.

Such cases can be approached using generalized linear models (GLMs). In this case, the variables deemed important by the scientist (that's you) can be included together in a linear model that is fit to the data, thus enabling estimation of expression changes based on those variables.


Data preparation and exploration
--------------------------------

We avoided such a situation in the example above by separating the glucose and ethanol conditions into two separate datasets, which were analyzed individually. But what if we wanted to combine these data together?

`r sn<-sn+1; paste(sn,". ", sep="")` Import both gene count files:
```{r}
# load the glucose data, and append column names with a "g"
glucose.data <- read.delim('data/gene_counts/gene_counts_glucose.txt',
                           stringsAsFactors=F, row.names=1)
colnames(glucose.data) <- paste(colnames(glucose.data), 'g', sep='')

# load the ethanol data, and append column names with a "e"
ethanol.data <- read.delim('data/gene_counts/gene_counts_ethanol.txt',
                           stringsAsFactors=F, row.names=1)
colnames(ethanol.data) <- paste(colnames(ethanol.data), 'e', sep='')
```


`r sn<-sn+1; paste(sn,". ", sep="")` Merge the count data by gene IDs (row names):
```{r}
# merge datasets (by=0 indicates we merge by row names)
count.data <- merge(glucose.data, ethanol.data, by=0)

# set the row names as the gene IDs, and remove the "Row.names" column
row.names(count.data) <- count.data$Row.names
count.data <- count.data[,-1]

# preview our merged dataframe
head(count.data)
```

`r sn<-sn+1; paste(sn,". ", sep="")` Create a `DGEList` object from the `count.data` dataframe:
```{r}
y <- DGEList(counts=count.data)
```

Notice that this time, we did not specify the sample groups when creating our `DGEList`. The group is only specified at this step when dealing with a single variable (as in the example above), which is no longer the case. We will specify the sample grouping assignments in the next step.

Take a look at the multidimensional scaling (MDS) plot (using colors this time):
```{r}
plotMDS(y, pch=16, 
        col=c(rep('red',6), rep('blue',24), rep('orange',6), rep('purple',15)))
legend(-2.5,-1.3, c('Pheromone Glc', 'Unaffected Glc', 'Pheromone EtOH', 'Unaffected EtOH'),
       col=c('red','blue','orange','purple'), pch=16, cex=0.9)
```

`r sn<-sn+1; paste(sn,". ", sep="")` Generate a design matrix to classify each of our samples based on carbon source and pheromone status:
```{r}
# specify pheromone and carbon attribute of each sample
pheromone <- factor(startsWith(colnames(count.data), 'P'), labels=c('unaffected','pheromone'))
carbon <- factor(endsWith(colnames(count.data), 'e'), labels=c('glucose','ethanol'))

# generate design matrix
design <- model.matrix(~carbon + pheromone)
rownames(design) <- colnames(y)
```

> **Question `r qn<-qn+1;qn`:** Does the design matrix make sense?

<div class="answer"><table class="answer"><tr><td class="answer">
**Answer:** 
```{r}
design[1:10, ]  # look at the first 10 rows
```

The ones and zeros indicate whether each sample is or is not (respectively) assigned to the group indicated in the column. Although the column names are a bit strange in `design`, we can see that sample `P1g` for example has a zero for `carbonethanol` because it is glucose, not ethanol, and a one for `pheromonepheromone` because it is pheromone-affected. 
</td></tr></table></div>

`r sn<-sn+1; paste(sn,". ", sep="")` Filter low-count genes from the count data, this time providing our design matrix because sample groups were not specified in the DGEList object:
```{r}
keep <- filterByExpr(y, design=design, min.count = 10)
y <- y[keep, , keep.lib.sizes=F]
```

`r sn<-sn+1; paste(sn,". ", sep="")` Calculate the normalization factors for each sample:
```{r}
y <- calcNormFactors(y)
```

`r sn<-sn+1; paste(sn,". ", sep="")` Estimate the dispersion across all genes:
```{r}
y <- estimateDisp(y, design)
```

`r sn<-sn+1; paste(sn,". ", sep="")` Take a look at the BCV plot:
```{r}
plotBCV(y)
```


We are ready to fit our GLM to the count data. To do this, we will use the `glmQLFit` function, which fits a quasi-likelihood negative binomial model to the count data.

`r sn<-sn+1; paste(sn,". ", sep="")` Fit the GLM:
```{r}
fit <- glmQLFit(y, design)
```

Alternatively, we could use the `glmFit` function, which uses a slightly different algorithm. You can find more details in the [edgeR User's Guide](https://bioconductor.org/packages/release/bioc/vignettes/edgeR/inst/doc/edgeRUsersGuide.pdf).

Now we want to use our fitted GLM to estimate the expression change between the pheromone-affected and pheromone-unaffected samples, while ignoring differences associated with the carbon source (glucose vs. ethanol). To do so, we will perfom what is known as "blocking", where the effect of the carbon source difference is "blocked" from our expression change estimation.

`r sn<-sn+1; paste(sn,". ", sep="")` Perform a likelihood test to estimate gene expression changes caused by the pheromone:
```{r}
qlf <- glmQLFTest(fit)
topTags(qlf)  # take a look at the most affected genes
```

Notice we didn't specify which comparison we wanted to perform (pheromone affected vs. unaffected, or glucose vs. ethanol). This is because by default the `glmQLFTest` function will test the last (3rd) coefficient in the model (`pheromone` in our case). If we instead wanted to look at gene expression changes between glucose and ethanol samples, we would specify this using the `coef` argument: e.g., `qlf <- glmQLFTest(fit, coeff=2)`, because the 2nd coefficient relates to carbon source (the 1st coefficient is simply the intercept, or baseline).


> **Question `r qn<-qn+1;qn`:** How do these top-affected genes compare to those in the first analysis?

<div class="answer"><table class="answer"><tr><td class="answer">
**Answer:** 
```{r}
# merge the results tables from both analyses
merged.results <- merge(qlf$table, et$table, by=0)

# compare the log-transformed p-values with a scatter plot
plot(-log10(merged.results$PValue.x), -log10(merged.results$PValue.y))
```

The results certainly show a similar trend, but it's also clear that the results have been substantially affected by our new analysis approach.
</td></tr></table></div>



Session info
============
This page was generated using the following R session:
```{r, echo=F}
sessionInfo()
```

