---
title: "Gene Set Analysis and Interpretation"
output: 
  html_document:
    toc: true
    toc_float:
      collapsed: false
    fig_width: 10
editor_options: 
  chunk_output_type: console
---

<style type="text/css">

blockquote {
  background: #ECF8FF;
  border-left: 10px solid #3989CB;
  margin: 1.5em 10px;
  padding: 0.5em 10px;
  font-size: 14px;
}

h1 { 
  font-size: 25px;
  margin-top: 1.5cm;
  margin-bottom: 0.5cm;
}
h2 { 
  font-size: 18px;
  margin-top: 1cm;
  margin-bottom: 0.5cm;
}
h3 {
  font-size: 14px;
  margin-top: 1cm;
  margin-bottom: 0.5cm;
}

table.answer, td.answer {
   border: 0px;
   background: #BCE0C0;
   padding: 10px;
   width: 100%;
}

div.answer { display: true;}

</style>


```{r setup, include=F}
knitr::opts_chunk$set(echo = TRUE, tidy=F, eval=T, cache=F)
qn <- sn <- 0
```

Setup and load data 
===================

The data needed for this exercise are the results from the previous differential expression analysis exercise, using the gene counts from glucose conditions (saved as `DE_results_glucose.txt` in the `data/DE_results/` subdirectory within the `IDA_tutorial` directory).

`r sn<-sn+1; paste(sn,". ", sep="")` Load the R packages that will be used in this exercise:
```{r, message=FALSE, warning=FALSE, results='hide'}
library(knitr)
library(piano)
library(pheatmap)
library(org.Sc.sgd.db)
library(edgeR)
```

If you are not able to load any of these packages, try to install them using the following commands:
```{r, eval=F}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install('PACKAGE_NAME_HERE')
```

Or alternatively:
```{r, eval=F}
install.packages('PACKAGE_NAME_HERE')
```

If something fails, try to understand the error message and fix it. If you get stuck, ask for help :)  

`r sn<-sn+1; paste(sn,". ", sep="")` Start by loading the differential expression results from the previous exercise:
```{r}
diffExpRes <- read.delim('data/DE_results/DE_results_glucose.txt', stringsAsFactors=F)
head(diffExpRes)  # take a quick look at our results dataframe
```


Web-based enrichment analysis
=============================

We will start by performing an overrepresentation analysis (a.k.a. list enrichment analysis, etc.) by using Enrichr and DAVID. Both use a scoring method similar to the Hypergeomtric/Fisher's exact test. To perform such a test, we need to have a list of interesting (e.g. differentially expressed) genes and a list of the background genes (also known as the "universe"). However, both Enrichr and DAVID have their own background lists, so we do not need to specify them explicitly. First, let's make a list of interesting genes!  

`r sn<-sn+1; paste(sn,". ", sep="")` You can use the `write.table` function to print the top significant genes to copy paste into some web browser tools:
```{r, results="hide"}
# here we arbitrarily chose genes with an FDR < 0.001 and absolute log2FC > 3
gene.index <- diffExpRes$FDR<0.001 & abs(diffExpRes$logFC)>3
selected.genes <- rownames(diffExpRes)[gene.index]
write.table(selected.genes, quote=F, row.names=F, col.names=F)
```


DAVID
-----

`r sn<-sn+1; paste(sn,". ", sep="")` Select and copy the gene list given by the above command and head over to [DAVID](https://david.ncifcrf.gov/).
`r sn<-sn+1; paste(sn,". ", sep="")` Locate *Functional Annotation* in the left menu and click the [more](https://david.ncifcrf.gov/helps/functional_annotation.html) link and briefly scroll through that page to get an overview of what the Functional Annotation Tool at DAVID does.

> **Question `r qn<-qn+1;qn`:** What is the alternative to the Fisher Exact P-Value used by DAVID?

<div class="answer"><table class="answer"><tr><td class="answer">
**Answer:** 

DAVID uses the so-called EASE Score. In the contingency table, 1 is subtracted from the number of genes that are both in the gene-list and in the gene-set.
</td></tr></table><br></div>

`r sn<-sn+1; paste(sn,". ", sep="")` Now go to the [*Functional Annotation* page](https://david.ncifcrf.gov/summary.jsp) and make sure the Upload tab is visible.
`r sn<-sn+1; paste(sn,". ", sep="")` Paste the copied gene-list, select the correct identifier ("ENSEMBL_GENE_ID"), and select the button specifying that this is a Gene List. Click *Submit List*.

> **Question `r qn<-qn+1;qn`:** Were all gene Ensembl IDs recognized? How many were unmapped?

<div class="answer"><table class="answer"><tr><td class="answer">
**Answer:** 
14 genes were unmapped, many of which were tRNA genes (e.g., "tR(ACG)L").
</td></tr></table><br></div> 

`r sn<-sn+1; paste(sn,". ", sep="")` Explore the results. For instance, click on Functional Annotation Clustering at the bottom of the page. This shows related gene-sets clustered together in groups for a nice overview.

> **Question `r qn<-qn+1;qn`:** What seems to be the main functions of the top significant genes (i.e. in summary what are the significantly enriched gene-sets)? Does this make sense considering the experimental conditions?

<div class="answer"><table class="answer"><tr><td class="answer">
**Answer:** 
(ADD ANSWER)
</td></tr></table><br></div>


Enrichr
-------

As an alternative to DAVID we can use [Enrichr](https://amp.pharm.mssm.edu/YeastEnrichr/). Note that there are Enrichr versions for other species, such as [Human](https://amp.pharm.mssm.edu/Enrichr/). This page takes gene symbols (names) as input, so first we need to translate our yeast Ensembl IDs to gene symbols. This can be done in many ways, one of which is to use the `org.Sc.sgd.db` package (find a user manual [here](https://bioconductor.org/packages/release/data/annotation/manuals/org.Sc.sgd.db/man/org.Sc.sgd.db.pdf)):

```{r}
# retrieve the org.Sc.sgd.db gene name mapping information
x <- org.Sc.sgdGENENAME

# determine which genes are mapped to a gene name (gene symbol)
mapped_genes <- mappedkeys(x)

# keep only entries with a name mapping
xmapped <- as.list(x[mapped_genes])

# retrieve names for our genes of interest
selected.gene.names <- xmapped[selected.genes]

# "flatten" the list into a vector, and remove empty entries
selected.gene.names <- unlist(selected.gene.names, use.names=F)
```


`r sn<-sn+1; paste(sn,". ", sep="")` Print this as a vector of gene names that can easily be copied:
```{r, results="hide"}
write.table(selected.gene.names, quote=F, row.names=F, col.names=F)
```

> **Question `r qn<-qn+1;qn`:** Were we able to map all of the gene Ensembl IDs to a gene name?

<div class="answer"><table class="answer"><tr><td class="answer">
**Answer:**
```{r}
length(selected.genes) - length(selected.gene.names)
```
No, many genes did not map, and therefore cannot be included in the analysis. In fact, nearly 25% of the genes in this case did not map, which is reason to treat the results with caution. In this case, one should look into alternative Enrichment tools or gene ID conversion approaches to reduce the number of genes that are "lost" in the process.
</td></tr></table><br></div> 


`r sn<-sn+1; paste(sn,". ", sep="")` Paste your list of gene names on the [Enrichr](https://amp.pharm.mssm.edu/YeastEnrichr/) website and click Submit.

`r sn<-sn+1; paste(sn,". ", sep="")` Explore some of the different gene set collections that are available. You can also navigate to different collection categories using the tabs near the top of the page ("Ontologies", "Pathways", "Transcription).

> **Question `r qn<-qn+1;qn`:** Are the results similar to those we got from DAVID?

<div class="answer"><table class="answer"><tr><td class="answer">
**Answer:** 
(ADD ANSWER)
</td></tr></table><br></div> 


> **Question `r qn<-qn+1;qn`:** For the Enrichr and DAVID results so far, can we know whether the identified functions are active/inactive or up/down-regulated in untreated vs pheromone-treated? If not, how can we obtain such information?

<div class="answer"><table class="answer"><tr><td class="answer">
**Answer:** 
The list we have used contains both up- and down-regulated genes, so we have not taken into account fold-change direction. One could subset the list to up- and down-regulated genes separately. Or use a method that takes directionality into account.
</td></tr></table><br></div> 



Enrichment analysis in R
========================


Piano
-----

Piano is an R-package for performing gene-set analysis (see more info [here](https://varemo.github.io/piano/)).

First we need to import some gene-sets. In this example, we will use gene sets retrieved from the [Saccharomyces Genome Database (SGD)](https://www.yeastgenome.org/). To make things easier, the gene sets have been pre-processed and saved as .RDS files which can be loaded directly into R. If you're interested in how the gene set file was pre-processed, the function can be found [here (NOTE: update this link to point directly at the function!)](https://github.com/JonathanRob/IDA_tutorial).

`r sn<-sn+1; paste(sn,". ", sep="")` Load one of the three available gene set collections (biological process, cellular component, or molecular function):
```{r}
gscGO <- readRDS('data/GSC/biological_process.rds')
# gscGO <- readRDS('data/GSC/cellular_component.rds')
# gscGO <- readRDS('data/GSC/molecular_function.rds')

# convert the dataframe to a GSC object
gscGO <- loadGSC(gscGO)
gscGO  # take a look
```

The printed info is always good to check. Did the genes get read as genes and the gene-sets as gene-sets? Do the gene-set sizes seem reasonable?

`r sn<-sn+1; paste(sn,". ", sep="")` Extract the gene-level statistics that we will use:

```{r}
# get geneIDs, pvals, and logfc:
geneIDs <- rownames(diffExpRes)

pvals <- diffExpRes$PValue
names(pvals) <- geneIDs

logfc <- diffExpRes$logFC
names(logfc) <- geneIDs

# to avoid issues with pvals of zero:
pvals[pvals == 0] <- min(pvals[pvals > 0])
```

`r sn<-sn+1; paste(sn,". ", sep="")` Inspect the objects we just created (e.g. using `head()`, `View()`, or `length()`) so that you know how they look and what they contain.  

We now have the gene-level data in a convenient format and are ready to continue with the analysis.

### Overrepresentation analysis

First, let's use piano to perform a hypergeometric test, similar to what we did with Enrichr and DAVID.  

`r sn<-sn+1; paste(sn,". ", sep="")` Run the following code and read the printed info to understand what it does:

```{r, message=F, warning=F, fig.height=6}
# Run the analysis:
res <- runGSAhyper(genes=selected.genes, universe=geneIDs, gsc=gscGO, gsSizeLim=c(10, Inf))

# Visualize results:
networkPlot2(res, class="non", adjusted=T, significance=0.05, ncharLabel=Inf,
            nodeSize=c(5,30), edgeWidth=c(2,10), labelSize=12, overlap=10, 
            scoreColors=c("greenyellow","darkgreen"))
```
Here, we use the `networkPlot2` function which draws a network for the most significant gene-sets (based on our provided input threshold, 0.05 in this case). The plot is also interactive - try dragging a node around!

> **Question `r qn<-qn+1;qn`:** What does `gsSizeLim=c(10, Inf)` in the code above mean?  

<div class="answer"><table class="answer"><tr><td class="answer">
**Answer:** 
To only include gene-sets with 10 or more genes in the analysis (after filtering out genes not in the expression data). This is usually necessary to remove relatively small and uninformative gene sets that contain one or very few genes.
</td></tr></table><br></div> 

> **Question `r qn<-qn+1;qn`:** What do the colors, node-sizes, and edges denote?  

<div class="answer"><table class="answer"><tr><td class="answer">
**Answer:** 
Node size corresponds to gene-set size, i.e. number of genes. Edges denote the number of shared genes between gene-sets. Note that gene-sets without edges can also share a few genes, depending on the value of the `overlap` parameter. The color corresponds to the significance (p-value).
</td></tr></table><br></div> 

> **Question `r qn<-qn+1;qn`:** Are the results in line with those from DAVID and Enrichr?

<div class="answer"><table class="answer"><tr><td class="answer">
**Answer:** 
(ADD ANSWER)
</td></tr></table><br></div> 

### Gene-set analysis

Next, we will use piano to perform gene-set analysis, i.e. not the enrichment of a gene-list based on a cut-off, but using all available gene-level statistics.

For the first example, we will use the GSEA (Gene Set Enrichment Analysis) method as described in [Subramanian et al. *PNAS* 2005](https://www.pnas.org/content/102/43/15545.abstract). The gene-level stats that we will use for this method will be the signed (to keep track of fold-change) -log10-pvalues, i.e. `-log10(pvals)*sign(logfc)`.

`r sn<-sn+1; paste(sn,". ", sep="")` Run the GSA using piano and plot the results:

```{r, message=F, warning=F, results='hide'}
gsaRes <- runGSA(-log10(pvals)*sign(logfc), geneSetStat="fgsea", gsc=gscGO, gsSizeLim=c(10, Inf))

networkPlot2(gsaRes, "distinct", "both", adjusted=T, significance=0.05, ncharLabel=Inf,
            nodeSize=c(5,30), edgeWidth=c(2,10), labelSize=12, overlap=10,
            scoreColors=c("red", "orange", "yellow", "blue", "lightblue", "lightgreen"))
```

`r sn<-sn+1; paste(sn,". ", sep="")` Read the printed info to check that the numbers make sense!

> **Question `r qn<-qn+1;qn`:** Look at the network plot, could gene overlap between your gene-sets bias the result interpretation in any way?  

<div class="answer"><table class="answer"><tr><td class="answer">
**Answer:** 
There are clusters of gene-sets that share a large portion of genes, i.e. they are in fact quite similar and in extreme cases basically measure the same thing. If these sets are many it is easy to focus the conclusions on the process that those sets represent, in contrast to the processes represented only by single or a few gene-sets. However, this maybe is due to the fact that one process is more fine-grained annotated by e.g. GO. The network plot helps in visualizing this, and one can try to see each cluster as an affected process, regardless of the number of gene-sets it contains.
</td></tr></table><br></div> 

The network plot can be nice for visualizing your results, but the function `GSAsummaryTable` can be used to get a table of the all the GSA results.

`r sn<-sn+1; paste(sn,". ", sep="")` Try it together with RStudios `View()` command:

```{r, eval=F}
View(GSAsummaryTable(gsaRes))
```

> **Question `r qn<-qn+1;qn`:** What is the top significant GO-term affected by down-regulation and up-regulation, respectively?  

<div class="answer"><table class="answer"><tr><td class="answer">
**Answer:** 
Up: ____
Down: _____
</td></tr></table><br></div> 


`r sn<-sn+1; paste(sn,". ", sep="")` Now run a second GSA, using Stouffer's method (option `geneSetStat="stouffer"`) instead of GSEA (`fgsea`):
```{r, message=F, warning=F, results='hide', echo=T}
gsaRes <- runGSA(pvals, sign(logfc), geneSetStat="stouffer", gsc=gscGO, gsSizeLim=c(20, 200))
```

In this case, we provide the gene p-values and fold-change direction (`sign(logfc)`) as separate inputs because this method does not require a single gene-level statistic like GSEA.


`r sn<-sn+1; paste(sn,". ", sep="")` Plot the results using a network plot, and play around with the options until you think it looks nice and presents the results from the GSA in a good way.

For example it could look something similar to this:
```{r, message=F, warning=F, echo=T, fig.height=6}
networkPlot2(gsaRes, "distinct", "both", adjusted=T, significance=0.001, ncharLabel=Inf,
             nodeSize=c(5,25), edgeWidth=c(2,10), labelSize=15, overlap=10, physics=F,
             scoreColors=c("red", "orange", "yellow", "blue", "lightblue", "lightgreen"))
```

> **Question `r qn<-qn+1;qn`:** Are these results in line with previous results that we have seen in this exercise? Do you feel there is a consistent pattern?

<div class="answer"><table class="answer"><tr><td class="answer">
**Answer:** 
(ADD ANSWER)
</td></tr></table><br></div> 

Another way to visualize the results instead of the network plot is using a heatmap. Piano's `GSAheatmap` function is an easy way to display the results as a clustered heatmap:
```{r, message=F, warning=F, results='hide', echo=T, fig.height=12}
GSAheatmap(gsaRes, adjusted=T, ncharLabel=Inf, cellnote='rank', cex=1.4)
```

Note that the values in the color key and shown in the heatmap cells are *rank* values, where 1 means it was the most significant gene set for that directionality class (column). There can of course be many ties for rank values, as we see in the current example.


<br>Yet _another_ way to view the results of a GSA (and the gene set collection itself) is through the Piano `exploreGSAresults` function, which uses rShiny to enable interactive exploration of your GSA results:
```{r, eval=F}
exploreGSAres(gsaRes)
```

This function will open up an interactive session in your internet browser. Simply close the tab/window when you're finished to continue working in your R session.  

<br>It is always good to go back to the gene-level while looking at your final GSA results. Here we will use a heatmap for visualizing gene-level data for selected gene-sets (using one of *many* heatmap functions available for R).

`r sn<-sn+1; paste(sn,". ", sep="")` Let's pull out the information we need for a specific gene-set:

```{r}
# Get genes for a specific gene-set:
selGenes <- names(geneSetSummary(gsaRes, "oligosaccharide metabolic process")$geneLevelStats)

# Merge with the diffExp result table:
selTab <- diffExpRes[rownames(diffExpRes)%in%selGenes,c(1,4)]
selTab$log10FDR <- -log10(selTab$FDR)*sign(selTab$logFC)

# Display this table (only showing the top 6 entries here):
kable(head(selTab))
```

`r sn<-sn+1; paste(sn,". ", sep="")` Now, let's fetch the count data for these genes and make a heatmap of all the information we have collected:

```{r, fig.height=6, dpi=600}
# Read in the count data:
count.data <- read.delim("data/gene_counts/gene_counts_glucose.txt",
                         stringsAsFactors=F, row.names=1)

# Merge count data with the selTab:
selTab <- merge(selTab, count.data, by=0, all.x=T)
row.names(selTab) <- selTab$Row.names

# Plot a heatmap
# Row annotation:
rowann <- data.frame(Significant=ifelse(selTab$FDR<0.05,"yes","no"),
                     Direction=ifelse(selTab$logFC>0,"Up","Down"))
row.names(rowann) <- row.names(selTab)
# Column annotation:
colann <- data.frame(Pheromone=ifelse(startsWith(colnames(selTab)[5:ncol(selTab)],'P'),
                                      'pheromone','untreated'))
row.names(colann) <- colnames(selTab[5:ncol(selTab)])
# Specify annotation colors:
ann_colors <- list(Significant = c(no='black', yes='green'),
                   Direction = c(Up='red', Down='blue'),
                   Pheromone = c(pheromone='yellow', untreated='black'))
# Specify row labels:
rowlabs <- selTab$Row.names
# Plot:
pheatmap(selTab[,5:ncol(selTab)], cluster_cols=F, scale="row",
         display_numbers=selTab[,5:ncol(selTab)], annotation_row=rowann, 
         annotation_col=colann, annotation_colors=ann_colors,
         labels_row=rowlabs)
```

Note that here we scale the rows so the colors will be visualizing the difference for each gene specifically, but is not comparable across genes. Therefore we also include the actual counts in each heatmap cell. Also, note that these are raw counts. As an alternative we could have plotted normalized CPM (counts per million) using the command:

```{r}
cpms <- cpm(calcNormFactors(DGEList(count.data), method='TMM'))
```

> **Question `r qn<-qn+1;qn`:** Why are the rows reordered?

<div class="answer"><table class="answer"><tr><td class="answer">
**Answer:** 
A clustering is performed on the rows to group genes with similar expression patterns across the different samples.
</td></tr></table><br></div> 


> **Question `r qn<-qn+1;qn`:** Given the count data for these genes, does the differential expression results make sense, as indicated by the column annotation columns? Any specific genes that you are concerned about?  

<div class="answer"><table class="answer"><tr><td class="answer">
**Answer:** 
These are the genes belonging to the "oligosaccharide metabolic process" gene set, which was shown to be affected by down-regulation in the GSA. Here we can see that a high portion of the genes are signifcant, and that the majority of significant genes are in fact down-regulated. Looking at the count data (the actual heatmap) one can also see a pattern of blue to the left (pheromone-treated) and red to the right (untreated), indicating a lower expression of these genes upon pheromone treatment.
</td></tr></table><br></div> 

> **Question `r qn<-qn+1;qn`:** Would you say from looking at the heatmap, that in general oligosaccharide metabolism is down-regulated or up-regulated? Does your answer reflect the GSA results as seen in the network plot/GSAheatmap above?

<div class="answer"><table class="answer"><tr><td class="answer">
**Answer:** 
Down-regulated. It does reflect the network plot results.
</td></tr></table><br></div> 

`r sn<-sn+1; paste(sn,". ", sep="")` Spot-check some of these genes to see what their functions are (Google search or similar).

> **Question `r qn<-qn+1;qn`:** Are these genes actually involved in oligosaccharide metabolism? Are you confident enough from the data and analysis results to say anything about how pheromone treatment affects this biological process?

<div class="answer"><table class="answer"><tr><td class="answer">
**Answer:** 
(ADD ANSWER)
</td></tr></table><br></div>


### Reporter Metabolite Analysis

In addition to gene ontology terms, we can also evaluate the enrichment of our DE results in the context of the metabolic network. One such approach is reporter metabolite analysis, as described in [Patil & Nielsen, *PNAS* 2005](https://www.pnas.org/content/102/8/2685). Reporter metabolite analysis is essentially the same as a typical gene-set analysis, but the gene sets are metabolites (e.g., NADH, glucose, citrate, etc.) instead of GO terms. Here we'll try out the reporter metabolite analysis using gene-metabolite associations derived from the yeast genome-scale metabolic model, [Yeast-GEM](https://github.com/SysBioChalmers/yeast-GEM).

`r sn<-sn+1; paste(sn,". ", sep="")` Load the Yeast-GEM network as a GSC:
```{r}
geneMetData <- read.delim('data/GSC/yeastGEM_gene2met.txt')
gscGO <- loadGSC(geneMetData)
gscGO  # take a look
```

*Note:* Normally we can load a GSC directly from a GEM .xml (SBML) file using `loadGSC`, but some larger genome-scale models (such as yeast-GEM) can take quite some time to load. Therefore, we have provided a pre-processed file `yeastGEM_gene2met.txt` which contains essentially the same information that would have been extracted from the SBML file.

`r sn<-sn+1; paste(sn,". ", sep="")` Run a reporter metabolite analysis using the yeast-GEM GSC:
```{r, message=F, warning=F}
gsaRes <- runGSA(pvals, sign(logfc), geneSetStat="reporter", gsc=gscGO, gsSizeLim=c(20, 200))
```

`r sn<-sn+1; paste(sn,". ", sep="")` Examine your results using `exploreGSAres`, `GSAheatmap`, or `networkPlot2`:
```{r, message=F, warning=F, fig.height=6}
# network plot example:
networkPlot2(gsaRes, "distinct", "both", adjusted=T, significance=0.05, ncharLabel=Inf,
             nodeSize=c(10,30), edgeWidth=c(3,10), labelSize=20, overlap=5, physics=T)
```


Session info
============
This page was generated using the following R session:
```{r, echo=F}
sessionInfo()
```

